src/eg/src/verifiable_decryption.rs

```diff
@@ -0,0 +1,717 @@
+        proof_commit_state: &DecryptionProofStateShare,
+        secret_key_share: &GuardianSecretKeyShare,
+    ) -> Result<DecryptionProofResponseShare, ResponseShareError> {
+        if proof_commit_state.i != secret_key_share.i {
+            return Err(ResponseShareError::KeyStateShareIndexMismatch {
+                i: proof_commit_state.i,
+                j: secret_key_share.i,
+            });
+        }
+
+        let p = fixed_parameters.p.as_ref();
+        let q = fixed_parameters.q.as_ref();
+
+        let (a, b) = proof_commit_shares
+            .iter()
+            .fold((BigUint::one(), BigUint::one()), |(a, b), share| {
+                ((a * share.a_i.clone()) % p, (b * share.b_i.clone()) % p)
+            });
+
+        let c = Self::challenge(h_e, k, ciphertext, &a, &b, m);
+        let i_big = BigUint::from(proof_commit_state.i.get_one_based_u32());
+        let xs: Vec<BigUint> = proof_commit_shares
+            .iter()
+            .map(|s| BigUint::from(s.i.get_one_based_u32()))
+            .collect();
+        let w_i = get_single_coefficient(&xs, &i_big, &fixed_parameters.q);
+        let c_i = (c * w_i) % q;
+
+        let v_i = fixed_parameters.q.subtract_group_elem(
+            &proof_commit_state.u_i,
+            &(c_i * secret_key_share.p_i.clone()),
+        );
+        Ok(DecryptionProofResponseShare {
+            i: proof_commit_state.i,
+            v_i,
+        })
+    }
+
+    /// This function computes a decryption proof given the commit shares and
+    /// The arguments are
+    /// - election_parameters - the election parameters
+    /// - h_e - the extended bash hash
+    /// - joint_key - the joint election public key
+    /// - ciphertext - the ciphertext
+    /// - decryption_shares - the decryption shares
+    /// - proof_commit_shares - the shares of the commit message
+    /// - proof_response_shares - the shares of the response
+    /// - guardian_public_keys - the guardians' public keys
+    /// This function checks that decryption_shares, proof_commit_shares,proof_response_shares, are *sorted* the same way, i.e. the i-th element in each list belongs to the same guardian. The function also checks that guardian_public_keys contains all n public keys.
+    pub fn combine_proof(
+        election_parameters: &ElectionParameters,
+        h_e: &HValue,
+        joint_key: &JointElectionPublicKey,
+        ciphertext: &Ciphertext,
+        decryption_shares: &[DecryptionShare],
+        proof_commit_shares: &[DecryptionProofCommitShare],
+        proof_response_shares: &[DecryptionProofResponseShare],
+        guardian_public_keys: &[GuardianPublicKey],
+    ) -> Result<Self, CombineProofError> {
+        let fixed_parameters = &election_parameters.fixed_parameters;
+
+        //Check that the indices match
+        for (ds, cs, rs) in izip!(
+            decryption_shares,
+            proof_commit_shares,
+            proof_response_shares
+        ) {
```

Comment:
Is it ensured somewhere else that these three lists have the same length?

Otherwise we should check it before zipping them.

Reply:
I have added a check to ensure that the lists are of the same length.